File: src/nios/cognitive/CognitiveOS.ts
/**
 * Cognitive OS — brainstem layer (Phase 0 stub).
 * Owns uncertainty, confidence, retries, decomposition, attention.
 */

import type { UILDocument } from "../contracts/types";

export interface CognitiveState {
  uncertainty: number;
  confidence: number;
  retryCount: number;
  attentionBudget: number;
  decompositionDepth: number;
}

export interface MetaDecision {
  action:
    | "retrieve"
    | "calculate"
    | "ask_user"
    | "simulate"
    | "debate"
    | "search"
    | "explain"
    | "optimize"
    | "execute_capability"
    | "escalate_model";
  reason: string;
  confidence: number;
}

const DEFAULT_STATE: CognitiveState = {
  uncertainty: 0.5,
  confidence: 0.5,
  retryCount: 0,
  attentionBudget: 4096,
  decompositionDepth: 0,
};

export class CognitiveOS {
  private state: CognitiveState = { ...DEFAULT_STATE };

  reset(): void {
    this.state = { ...DEFAULT_STATE };
  }

  getState(): CognitiveState {
    return { ...this.state };
  }

  updateFromUIL(uil: UILDocument): void {
    this.state.confidence = uil.confidence;
    this.state.uncertainty = Math.max(0, 1 - uil.confidence);
  }

  shouldRetry(maxRetries = 2): boolean {
    return this.state.retryCount < maxRetries && this.state.uncertainty > 0.4;
  }

  recordRetry(): void {
    this.state.retryCount += 1;
  }

  allocateAttention(tokens: number): boolean {
    if (tokens > this.state.attentionBudget) return false;
    this.state.attentionBudget -= tokens;
    return true;
  }

  metaDecide(uil: UILDocument, availableCapabilities: string[]): MetaDecision {
    const text = (uil.source_text || "").toLowerCase();

    if (/\b(balance|bakaya|baki|शेष)\b/.test(text)) {
      return {
        action: "execute_capability",
        reason: "Ledger balance query — use deterministic ERP capability",
        confidence: 0.95,
      };
    }

    if (/\b(vat|tds|tax|कर|भ्याट)\b/.test(text) && /\d/.test(text)) {
      return {
        action: "calculate",
        reason: "Numeric tax query — deterministic engine required",
        confidence: 0.92,
      };
    }

    if (/\b(what if|yadi|bhaye|simulate|optimize|best)\b/.test(text)) {
      return {
        action: "simulate",
        reason: "Counterfactual or optimization intent detected",
        confidence: 0.8,
      };
    }

    if (uil.confidence < 0.6) {
      return {
        action: "ask_user",
        reason: "Low UIL confidence — clarification needed",
        confidence: 0.7,
      };
    }

    if (availableCapabilities.some((c) => c.includes("knowledge"))) {
      return {
        action: "retrieve",
        reason: "Default — retrieve knowledge before reasoning",
        confidence: 0.75,
      };
    }

    return {
      action: "escalate_model",
      reason: "No deterministic path — escalate to reasoner",
      confidence: 0.6,
    };
  }
}

export const cognitiveOS = new CognitiveOS();

File: erp_bot/src/nios/cognitive/cognitive_os.py
"""Cognitive OS — uncertainty, retries, decomposition, attention."""

from __future__ import annotations

from dataclasses import dataclass, field

from .attention import AttentionBudget, allocate_attention
from .retry import RetryDecision, decide_retry
from .uncertainty import UncertaintyState, estimate_uncertainty


@dataclass
class MetaDecision:
    action: str
    reason: str
    confidence: float
    uncertainty: float = 0.0
    retry: RetryDecision | None = None
    attention: AttentionBudget | None = None
    sub_problems: list[str] = field(default_factory=list)


class CognitiveOS:
    """Brainstem between kernel and planner — owns retries and attention."""

    def meta_decide(
        self,
        text: str,
        uil_confidence: float,
        capabilities: list[str],
        *,
        attempt: int = 0,
        evidence_coverage: float = 1.0,
        token_budget: int = 4000,
    ) -> MetaDecision:
        uncertainty = estimate_uncertainty(text, uil_confidence, evidence_coverage)
        attention = allocate_attention(token_budget, text, capabilities)
        retry = decide_retry(uncertainty.score, attempt, evidence_coverage)
        sub_problems = self.decompose(text, uil_confidence)

        lower = text.lower()
        import re

        if re.search(r"\b(balance|bakaya|baki|शेष)\b", lower, re.I):
            return MetaDecision(
                "execute_capability", "Ledger balance — deterministic path", 0.95,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        if re.search(r"\b(vat|tds|tax|कर)\b", lower, re.I) and re.search(r"\d", text):
            return MetaDecision(
                "calculate", "Numeric tax — engine required", 0.92,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        if re.search(r"\b(what if|yadi|simulate|optimize|best)\b", lower, re.I):
            return MetaDecision(
                "simulate", "Counterfactual intent", 0.8,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        if retry.should_retry:
            return MetaDecision(
                "retrieve", retry.reason, 0.7,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        if uil_confidence < 0.6 or uncertainty.score > 0.5:
            return MetaDecision(
                "ask_user", "Low confidence — clarify", 0.7,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        if any("knowledge" in c for c in capabilities):
            return MetaDecision(
                "retrieve", "Retrieve before reason", 0.75,
                uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
            )

        return MetaDecision(
            "escalate_model", "No deterministic path", 0.6,
            uncertainty=uncertainty.score, retry=retry, attention=attention, sub_problems=sub_problems,
        )

    def decompose(self, text: str, uil_confidence: float) -> list[str]:
        """Break ambiguous requests into UIL sub-problems."""
        import re

        parts: list[str] = []
        if " and " in text.lower():
            parts = [p.strip() for p in re.split(r"\band\b", text, flags=re.I) if p.strip()]
        elif uil_confidence < 0.65 and len(text.split()) > 8:
            parts = [text[: len(text) // 2], text[len(text) // 2 :]]
        return parts[:3]


cognitive_os = CognitiveOS()
